Optimize numerical operations with python integer variable

matusvalo opened the pull request:
For now let's get optimisation of int variable on par with untype variable. Further optimisations can be done in separate PR.

Fixes #5783.

matusvalo commented:
For now marking as draft. Should I add tests for that? I was not able to find any tests for numerical operations with python int type so I suppose yes... Please let me know.

da-woods commented:
Last time I changed something relating to these optimisations (https://github.com/cython/cython/pull/4821/files) I put a test in tests/run/pyintop.pyx

da-woods commented on Cython/Compiler/Optimize.py:
```diff
@@ -3463,6 +3463,20 @@ def _handle_simple_method_object___truediv__(self, node, function, args, is_unbo
+    _handle_simple_method_int___mod__ = _handle_simple_method_object___mod__
+    _handle_simple_method_int___floordiv__ = _handle_simple_method_object___floordiv__
+    _handle_simple_method_int___truediv__ = _handle_simple_method_object___truediv__
+
```
Just curious about why you removed `__div__` in the second commit? (Not suggesting you should spend time making it work if it doesn't though... just asking what doesn't work)

matusvalo replied:
I added it first but then removed because:
1. it does not optimisation, if I am not wrong
2. it is only for python2 so I decided to simplify the PR

matusvalo replied:
Please let me know if you think it should be present though I would added but for me it is not worth it.

da-woods replied:
No that seems fine. I think we can skip this for Python 2. I hadn't realised they'd changed the names in Python 3.

matusvalo commented:
> Last time I changed something relating to these optimisations (https://github.com/cython/cython/pull/4821/files) I put a test in tests/run/pyintop.pyx

OK I added tests there. I did not check anything special, only that execution of code produces expected results. Not sure if there is a way of more deep tests...

da-woods commented on tests/run/pyintop.pyx:
```diff
@@ -515,3 +515,130 @@ def test_avoid_if_coercion(obj):
     else:
         return False
+
+def pure_add_int(obj2: int):
```
Probably worth adding `test_fail_if_path_exists`/`test_assert_path_exists`. I'd guess it should fail with `BinopNode`, and we should require a `CApiCallNode`. But I'm not completely sure of this off the top of my head.

For debugging purposes it's useful to temporarily add `PrintTree()` to the end of the pipeline in `Pipeline.py` - I think there might be a command line argument to do it but I find it easier just to edit the code

matusvalo replied:
> Probably worth adding test_fail_if_path_exists/test_assert_path_exists. I'd guess it should fail with BinopNode, and we should require a CApiCallNode

I was thinking about that. I did not added them because the PR is not changing tree but I will try to add them.

matusvalo replied:
> For debugging purposes it's useful to temporarily add PrintTree()

I suppose this is just a note, It should not be added in this PR...

da-woods replied:
I believe the PR will change the tree - it replaces a `BinopNode` with a function call of the optimized function.

> I  suppose this is just a note, It should not be added in this PR...

Yes - it was just a hint about how to work out what the tree is changed to.

matusvalo replied:
> I believe the PR will change the tree
Hmm OK I was not aware of that. I am still stranger in Cython internals...

matusvalo replied:
OK done. It was good hint because I found out that PR was not exercised because annotations was ignored. Hence, there are two major changes:
1. language_level was set to 3, I hope it does not affect rest of the tests
2. I have removed tests for non-Commutativity operations because it generates different code when operators are reversed.

da-woods commented on tests/run/pyintop.pyx:
```diff
@@ -515,3 +515,145 @@ def test_avoid_if_coercion(obj):
+
+import sys
+
+if sys.version_info < (3, 0):
```
It might just be easier to drop this test rather than spending too much time working out how to make it work in Python 2. (I think __all__ docstrings might be put into a tests dictionary on Python 2 at compile time so this might be hard to make work right)

matusvalo replied:
I think I fixed it. In dirty way but hopefully it will be deleted soon when we will clean cython from all python 2.7 old crust.

matusvalo replied:
This will allow to have code exercised in python3.

da-woods reviewed (approved):
Looks good to me I think

scoder commented on tests/run/pyintop.pyx:
```diff
@@ -1,3 +1,4 @@
+# cython: language_level=3
```
Let's just use the new default. No need to nail the string  type as well.
```suggestion
# cython: language_level=3str
```

scoder commented on tests/run/pyintop.pyx:
```diff
@@ -515,3 +516,158 @@ def test_avoid_if_coercion(obj):
     else:
         return False
+
+@cython.test_fail_if_path_exists('//AddNode')
```
I'd really check for a generic `BinopNode` for the tests, not a specific one. Especially not something as half-specific as `IntBinopNode`. Whether that's used or not depends on other factors than just this PR. We're not expecting any `BinopNode` at all in the test functions and don't want the tests to become no-ops by accident when we add other optimisations.

matusvalo replied:
> I'd really check for a generic `BinopNode` for the tests, not a specific one. 

When trying to substitude `'//AddNode'` with `'//BinopNode'` the test is not failing when the fix is not present. I suppose tests added should fail when the fix is not present (e.g. on master branch). Moreover, when I print the tree of the test there is no `BinopNode` present:
```
          - stats[0]: DefNode(name="pure_add_int")
            - args[0]: CArgDeclNode(pos=(pyintop.pyx:522:17))
              - base_type: CAnalysedBaseTypeNode(pos=(pyintop.pyx:522:23))
              - declarator: CNameDeclaratorNode(pos=(pyintop.pyx:522:17))
              - annotation: AnnotationNode(type=None)
            - body: StatListNode(pos=(pyintop.pyx:523:4))
              - stats[0]: CompilerDirectivesNode(pos=(pyintop.pyx:523:4))
                - body: StatListNode(pos=(pyintop.pyx:523:4))
                  - stats[0]: SingleAssignmentNode(pos=(pyintop.pyx:527:16))
                    - lhs: NameNode(type=<PyObjectType>, name="res1")
                    - rhs: AddNode(type=<PyObjectType>)
                      - operand1: NameNode(type=<PyInt_Type>, name="obj2")
                      - operand2: IntNode(type=<PyObjectType>, value='1')
```
That's why I used specific nodes instead of generic `BinopNode`. Am I missing something?

da-woods replied:
Don't think so. At least according to the wiki (https://github.com/cython/cython/wiki/HackerGuide#parse-tree-assertions)

> You can express assertions using a simple XPath-like language called TreePath that traverses the parse tree. Nodes are referred to by their type name (inheritance is not considered).

It's possible we might want to improve that but I don't think as part of this PR.

-----------------------

One thing you could do is add separate "fail" tests for the base classes. E.g. this one would be 

```
@cython.test_fail_if_path_exists('//AddNode')
@cython.test_fail_if_path_exists('//NumBinopNode')
@cython.test_fail_if_path_exists('//BinopNode')
```

That gives a higher chance of catching it if some optimization changes

matusvalo replied:
Fixed as proposed by @da-woods .

scoder commented:
This is essentially a new feature and should not go into 3.0. It can go into 3.1.

matusvalo commented:
> This is essentially a new feature and should not go into 3.0. It can go into 3.1.

Fair enough. I thought that change is small so we can put it even in 3.0.X but I have no problem to move it to 3.1.

matusvalo commented:
Please let me know if there is any opposition with this PR. If not I will merge it soon.

da-woods commented:
> Please let me know if there is any opposition with this PR. If not I will merge it soon.

I think merge it into the master branch. Given we're treating 3.0.x as bugfix only, I don't think this justifies a backport.

matusvalo commented:
> > Please let me know if there is any opposition with this PR. If not I will merge it soon.
> 
> I think merge it into the master branch. Given we're treating 3.0.x as bugfix only, I don't think this justifies a backport.

Yes the target branch is master and version is 3.1.
